EXTENSIONS.md

```diff
@@ -0,0 +1,236 @@
+
+Annotation semantics may be useful for cases like ORMs and serialization frameworks, which need information about class fields, without affecting their normal runtime semantics. However, the popular ecosystem examples that we've found of this form, using just metadata for fields, seem to depend on metadata generated by TypeScript types. From these examples, it seems that annotation syntax alone would not be a sufficient solution.
+
+Some frameworks, including Angular, tend to use decorators which operate primarily by adding metadata. However, object literal annotations are not quite suitable for this usage, as they don't provide a way to be annotated in TypeScript to check types the way that functions do, they don't allow any processing of the metadata in code before it is saved, and because they don't provide a usable, stable identifier to be used for custom static analysis tools like tree shaking. For Angular, it may make more sense to use decorators which add metadata.
+
+For these reasons, annotations are omitted from this proposal's "MVP" (minimum viable product) and considered as a possible follow-on proposal.
+
+## Function decorators and annotations
+
+The `@logged` decorator from earlier would Just Work(TM) on a function, with function decorators!
+
+```js
+@logged
+@{x: "y"}
+function f() { }
+
+f();                        // prints logging information
+f[Symbol.annotations][0].x  // "y"
+```
+
+Function declarations with decorators or annotations are not hoisted. This is because it would be unintuitive to reorder the evaluation of the decorator or annotation expressions.
+
+Instead, functions with decorators or annotations are defined only when their declaration is reached. If they are used before they are defined, a ReferenceError is thrown, like classes. This ReferenceError condition is sometimes referred to as a "temporal dead zone" (TDZ). The TDZ risks unfortunate situations when refactoring, but at least those situations lead to easy-to-debug errors rather than the wrong function being run.
+
+Function decorator details:
+- First parameter: the function being decorated (or, whatever the next inner decorator returned)
+- Second parameter: a context object which just has `{ kind: 'function' }`
+- Return value: a new function, or undefined to keep the same function
+
+The inner binding of a function expression inside itself is in TDZ until all the function decorators run.
+
+## Parameter decorators and annotations
+
+A parameter decorator wraps the value of a function/method argument. It returns a function which does the wrapping.
+
+```js
+function dec(_, context) {
+  assert(context.kind === "parameter");
+  return arg => arg - 1;
+}
+
+function f(@dec @{x: "y"} arg) { return arg * 2 ; }
```

Comment:
if the order of these is swapped, does the decorator see the existing annotations? or are the annotations all queued up and applied to the final result, meaning that decorator-space and annotation-space don't interleave?

Reply:
The annotations are all queued up and applied to the final result (there's just no API at all for a decorator to see other annotations). The order is that the inner things apply first, and then can be overridden by outer things.